Swerve Drive and Positioning

dslit3 opened the pull request:
- Ports swerve drive and positioning from 2023 with some changes

dslit3 committed:
Initial port of swerve from 2023

- Many now-deprecated classes whose alternatives should be found before production
- Does not currently pass Spotless

dslit3 committed:
Apply spotless

jordan-powers commented:
We have spotless this year because it's easy (just run `./gradlew spotlessApply` before committing), and it makes reviews much nicer to have a unified coding style.

jordan-powers commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -4,15 +4,74 @@
+    private final NetworkButton calibrateTurnButton;
+
     public RobotContainer() {
+        ;
```
Typo

jordan-powers commented on src/main/java/frc/robot/command/CalibrateSwerveDriveCommand.java:
```diff
@@ -0,0 +1,256 @@
+import java.util.ArrayList;
+import java.util.function.DoubleSupplier;
+
+public class CalibrateSwerveDriveCommand extends Command {
```
At some point, we should probably move this class (and other swerve classes, such as SwerveModule) to MoUtils.
That doesn't need to happen now, though. Maybe offseason?

jordan-powers commented on src/main/java/frc/robot/math/Vec2.java:
```diff
@@ -0,0 +1,66 @@
+import org.ejml.simple.SimpleMatrix;
+
+/** Vector in 2 dimensions */
+public class Vec2 extends Vector<N2> {
```
What's this class for? Just a convenience wrapper for Vector<N2>?

dslit3 replied:
Replaces MoveVector (which represented fwd and left request, and handled applying curves), and because it implements the ejml vector wrapper we can do any other kind of math more easily with it

jordan-powers commented on src/main/java/frc/robot/subsystem/DriveSubsystem.java:
```diff
@@ -0,0 +1,272 @@
+        rearRight.drive(states[3]);
+    }
+
+    public boolean isMoving() {
+        return (frontLeft.driveMotor.getSelectedSensorVelocity() / MoPrefs.flDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF
+                || (frontRight.driveMotor.getSelectedSensorVelocity()
+                                / MoPrefs.frDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF
+                || (rearLeft.driveMotor.getSelectedSensorVelocity() / MoPrefs.rlDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF
+                || (rearRight.driveMotor.getSelectedSensorVelocity()
+                                / MoPrefs.rrDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF;
+    }
```
Do we use this function?

dslit3 replied:
Yes, we do not accept any april tag positioning if the robot is moving past the cutoff rate

jordan-powers commented on src/main/java/frc/robot/subsystem/PositioningSubsystem.java:
```diff
@@ -0,0 +1,173 @@
+     * The size of the field, in meters. Used to transform alliance-relative coordinates into
+     * field-relative coordinates.
+     */
+    private static final Translation2d fieldSize = new Translation2d(16.54175, 8.0137);
```
Is this still valid for this year's field?

jordan-powers commented on src/main/java/frc/robot/util/MoPrefs.java:
```diff
@@ -0,0 +1,120 @@
+
+/** Robot preferences, accessible through Shuffleboard */
+public class MoPrefs {
+    public static final Pref<Double> maxDriveSpeed = doublePref("Drive Max Speed m/s", 6000.0);
+    public static final Pref<Double> maxTurnSpeed = doublePref("Turn Max Speed m/s", 6000.0);
```
It has always bothered me that these preferences have slashes (`/`) in their name, since it causes the shuffleboard to treat them as a tree of preferences. But I never got around to renaming them, since I didn't want to lose the saved values.

Let's take the opportunity to rename these variables now.

jordan-powers replied:
Maybe we could add some logic to the constructor of Pref to reject/fix names that contain slashes?

jordan-powers commented on src/main/java/frc/robot/util/PoseFilter.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.Optional;
+
+/** Filters a stream of Pose3d to remove outliers and potentially bad data. */
+public class PoseFilter {
```
This could also live in MoUtils. Something else for the off-season.

jordan-powers reviewed (changes requested):
Looks mostly fine. The only thing I really need changed before I merge is the names of the `maxDriveSpeed` and `maxTurnSpeed` preferences.

jordan-powers committed:
Add more sim files to .gitignore

jordan-powers committed:
Rename MoPrefs with slash in key

dslit3 committed:
Upgrade to Phoenix v6 and new Spark Max classes

- Transitions away from deprecated libraries
- TalonFX PID no longer supports IZone

dslit3 committed:
Merge branch 'swerve-port' of https://github.com/momentumfrc/2024Crescendo into swerve-port

jordan-powers committed:
Update field size for 2024 field

dslit3 commented:
CTRE Phoenix v5 has been removed and I have done my best to migrate all code managing Talon FX motor controllers to use the CTRE Phoenix v6 versions, would be worth a second look-over

jordan-powers commented on src/main/java/frc/robot/util/MoTalonFxPID.java:
```diff
@@ -16,39 +21,55 @@ public class MoTalonFxPID {
 
     public void setIZone(double iZone) {
-        motorController.config_IntegralZone(0, iZone);
+        // TODO: This appears to no longer be supported by Phoenix v6
```
While this is unfortunate, there's nothing you can do about it. Luckily, we don't currently use the TalonFXs for position control.
